ext/msgpack/packer_ext_registry.c

```diff
@@ -40,24 +40,24 @@ void msgpack_packer_ext_registry_mark(msgpack_packer_ext_registry_t* pkrg)
     rb_gc_mark(pkrg->cache);
 }
 
-void msgpack_packer_ext_registry_dup(msgpack_packer_ext_registry_t* src,
+void msgpack_packer_ext_registry_dup(VALUE owner, msgpack_packer_ext_registry_t* src,
         msgpack_packer_ext_registry_t* dst)
 {
     if(RTEST(src->hash) && !rb_obj_frozen_p(src->hash)) {
-        dst->hash = rb_hash_dup(src->hash);
-        dst->cache = RTEST(src->cache) ? rb_hash_dup(src->cache) : Qnil;
+        RB_OBJ_WRITE(owner, &dst->hash, rb_hash_dup(src->hash));
+        RB_OBJ_WRITE(owner, &dst->cache, rb_hash_dup(src->cache));
     } else {
         // If the type registry is frozen we can safely share it, and share the cache as well.
-        dst->hash = src->hash;
-        dst->cache = src->cache;
+        RB_OBJ_WRITE(owner, &dst->hash, src->hash);
+        RB_OBJ_WRITE(owner, &dst->cache, src->cache);
```

Comment:
Is this cache always safe to share in this case now? There are now cases where `!RTEST(src->hash) && RTEST(src->cache)` (when the object has just been created in `msgpack_packer_ext_registry_init`).

Reply:
So if `!RTEST(src->hash)` is true, the `cache` just isn't used at all. But let see if we can make this better by setting it to `Qnil` in that case.

Reply:
I changed the code to be more explicit.

Reply:
Right, but if we dup a newly created `MessagePack::Factory`, then it will not have a `hash` but have a `cache`. The `hash` is lazily created in the first call to `MessagePack::Factory#register_type` (in `msgpack_packer_ext_registry_put`), which means that two `MessagePack::Factory` could share a `cache` but have completely different `hash`.

Reply:
Ah I see. Yeah, the issue is that this function is used for `Factory#dup` in which case it should dup both `hash` and `cache` given that it should return a mutable copy of the factory.

I only has the `Factory#packer` callsite in mind. I think it will be simpler to split the two use cases.